File: project/backend/app/models/vitals.py
"""
Vitals model for patient vital sign recordings
"""
from datetime import datetime

from sqlalchemy import Column, DateTime, Float, ForeignKey, Integer, Text
from sqlalchemy.orm import relationship

from app.db.base import BaseModel


class Vitals(BaseModel):
    """
    Stores a single set of vital sign measurements for a patient.

    BMI is computed in the application layer (POST /vitals/) from weight_kg
    and height_cm and stored here for fast retrieval.  recorded_at tracks
    when the measurement was physically taken; created_at (inherited from
    BaseModel) tracks when the row was inserted.

    Attributes:
        patient_id:         FK -> patients.id
        appointment_id:     FK -> appointments.id (nullable)
        recorded_by:        FK -> users.id (nurse or doctor who entered the data)
        recorded_at:        When the measurement was physically taken
        systolic_bp:        Systolic blood pressure in mmHg
        diastolic_bp:       Diastolic blood pressure in mmHg
        heart_rate:         Heart rate in bpm
        temperature:        Body temperature in degrees Fahrenheit
        respiratory_rate:   Respiratory rate in breaths per minute
        oxygen_saturation:  SpO2 as a percentage (0-100)
        weight_kg:          Body weight in kilograms
        height_cm:          Height in centimetres
        bmi:                Body Mass Index (weight_kg / (height_cm/100)^2)
        notes:              Free-text clinical notes
    """
    __tablename__ = "vitals"

    patient_id = Column(
        Integer, ForeignKey("patients.id", ondelete="CASCADE"),
        nullable=False, index=True,
    )
    appointment_id = Column(
        Integer, ForeignKey("appointments.id", ondelete="SET NULL"),
        nullable=True, index=True,
    )
    recorded_by = Column(
        Integer, ForeignKey("users.id", ondelete="SET NULL"),
        nullable=False, index=True,
    )
    recorded_at = Column(
        DateTime, nullable=False, default=datetime.utcnow, index=True,
    )

    systolic_bp = Column(Integer, nullable=True)
    diastolic_bp = Column(Integer, nullable=True)
    heart_rate = Column(Integer, nullable=True)
    temperature = Column(Float, nullable=True)
    respiratory_rate = Column(Integer, nullable=True)
    oxygen_saturation = Column(Float, nullable=True)
    weight_kg = Column(Float, nullable=True)
    height_cm = Column(Float, nullable=True)
    bmi = Column(Float, nullable=True)
    notes = Column(Text, nullable=True)

    # Relationships (one-directional — no back_populates on the other side
    # so existing models are not modified)
    patient = relationship("Patient", foreign_keys=[patient_id])
    recorder = relationship("User", foreign_keys=[recorded_by])
    appointment = relationship("Appointment", foreign_keys=[appointment_id])

    def __repr__(self) -> str:
        return f"<Vitals(id={self.id}, patient_id={self.patient_id}, recorded_at={self.recorded_at})>"

File: project/mobile/lib/models/vitals.dart
/// Vitals model representing a single set of vital sign measurements
class Vitals {
  final int id;
  final int patientId;
  final int? appointmentId;
  final int recordedBy;
  final String? recordedByName;
  final DateTime recordedAt;
  final int? systolicBp;
  final int? diastolicBp;
  final int? heartRate;
  final double? temperature;
  final int? respiratoryRate;
  final double? oxygenSaturation;
  final double? weightKg;
  final double? heightCm;
  final double? bmi;
  final String? notes;
  final DateTime createdAt;

  Vitals({
    required this.id,
    required this.patientId,
    this.appointmentId,
    required this.recordedBy,
    this.recordedByName,
    required this.recordedAt,
    this.systolicBp,
    this.diastolicBp,
    this.heartRate,
    this.temperature,
    this.respiratoryRate,
    this.oxygenSaturation,
    this.weightKg,
    this.heightCm,
    this.bmi,
    this.notes,
    required this.createdAt,
  });

  factory Vitals.fromJson(Map<String, dynamic> json) {
    return Vitals(
      id: json['id'] as int,
      patientId: json['patient_id'] as int,
      appointmentId: json['appointment_id'] as int?,
      recordedBy: json['recorded_by'] as int,
      recordedByName: json['recorded_by_name'] as String?,
      recordedAt: DateTime.parse(json['recorded_at'] as String),
      systolicBp: json['systolic_bp'] as int?,
      diastolicBp: json['diastolic_bp'] as int?,
      heartRate: json['heart_rate'] as int?,
      temperature: (json['temperature'] as num?)?.toDouble(),
      respiratoryRate: json['respiratory_rate'] as int?,
      oxygenSaturation: (json['oxygen_saturation'] as num?)?.toDouble(),
      weightKg: (json['weight_kg'] as num?)?.toDouble(),
      heightCm: (json['height_cm'] as num?)?.toDouble(),
      bmi: (json['bmi'] as num?)?.toDouble(),
      notes: json['notes'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'patient_id': patientId,
      'appointment_id': appointmentId,
      'recorded_by': recordedBy,
      'recorded_by_name': recordedByName,
      'recorded_at': recordedAt.toIso8601String(),
      'systolic_bp': systolicBp,
      'diastolic_bp': diastolicBp,
      'heart_rate': heartRate,
      'temperature': temperature,
      'respiratory_rate': respiratoryRate,
      'oxygen_saturation': oxygenSaturation,
      'weight_kg': weightKg,
      'height_cm': heightCm,
      'bmi': bmi,
      'notes': notes,
      'created_at': createdAt.toIso8601String(),
    };
  }

  /// True if systolic > 140 or diastolic > 90 (hypertension threshold)
  bool get isBpElevated =>
      (systolicBp != null && systolicBp! > 140) ||
      (diastolicBp != null && diastolicBp! > 90);

  /// True if heart rate > 100 bpm (tachycardia) or < 50 bpm (bradycardia)
  bool get isHeartRateAbnormal =>
      heartRate != null && (heartRate! > 100 || heartRate! < 50);

  /// True if SpO2 < 95%
  bool get isSpO2Low => oxygenSaturation != null && oxygenSaturation! < 95;

  /// True if temperature > 99.5°F (low-grade fever threshold)
  bool get isTempElevated => temperature != null && temperature! > 99.5;
}
